File: populatingNextRightII.py
import collections

class Node:
    def __init__(self, val: int = 0, left: 'Node' = None, right: 'Node' = None, next: 'Node' = None):
        self.val = val
        self.left = left
        self.right = right
        self.next = next

class Solution:
    def connect(self, root: 'Node') -> 'Node':
        if root == None:
            return root
        
        Q = collections.deque([root])
        while Q:
            size = len(Q)
            prev = None
            for i in range(size):
                curr = Q.popleft()
                if prev:
                    prev.next = curr
                prev = curr
                if curr.left:
                    Q.append(curr.left)
                if curr.right:
                    Q.append(curr.right)
        
        return root

File: populatingNextRightII.java
import java.util.LinkedList;
import java.util.Queue;

class Node {
    public int val;
    public Node left;
    public Node right;
    public Node next;

    public Node() {}
    
    public Node(int _val) {
        val = _val;
    }

    public Node(int _val, Node _left, Node _right, Node _next) {
        val = _val;
        left = _left;
        right = _right;
        next = _next;
    }
}

public class populatingNextRightII {
    // 117. Populating Next Right Pointers in Each Node II
    public Node connect(Node root) {
        if (root == null) return root;
        Node ret = root;
        Queue<Node> queue = new LinkedList<>();
        queue.offer(root);

        while (!queue.isEmpty()) {
            int size = queue.size();
            Node prev = null;
            while (size-- > 0) {
                Node curr = queue.poll();
                if (prev != null) {
                    prev.next = curr;
                }
                prev = curr;
                if (curr.left != null) queue.offer(curr.left);
                if (curr.right != null) queue.offer(curr.right);
            }
            prev.next = null;
        }
        
        return ret;
    }
}
